packages/core/utils/src/caching/index.ts

```diff
@@ -0,0 +1,238 @@
+   */
+  ttl?:
+    | number
+    | ((args: TargetMethodArgs<Target, PropertyKey>) => number | undefined)
+  /**
+   * Whether to auto invalidate the cache whenever it is possible.
+   */
+  autoInvalidate?:
+    | boolean
+    | ((args: TargetMethodArgs<Target, PropertyKey>) => boolean | undefined)
+  /**
+   * The providers to use for the cache.
+   */
+  providers?:
+    | string[]
+    | ((args: TargetMethodArgs<Target, PropertyKey>) => string[] | undefined)
+
+  container: MedusaContainer | ((this: Target) => MedusaContainer)
+}) {
+  return function (
+    target: Target,
+    propertyKey: PropertyKey,
+    descriptor: PropertyDescriptor
+  ) {
+    const originalMethod = descriptor.value
+
+    if (typeof originalMethod !== "function") {
+      throw new Error("@cached can only be applied to methods")
+    }
+
+    descriptor.value = async function (
+      ...args: Target[PropertyKey & keyof Target] extends (
+        ...args: any[]
+      ) => any
+        ? Parameters<Target[PropertyKey & keyof Target]>
+        : never
+    ) {
+      const container: MedusaContainer =
+        typeof options.container === "function"
+          ? options.container.call(this)
+          : options.container
+
+      const cachingModule = container.resolve<ICachingModuleService>(
+        Modules.CACHING,
+        {
+          allowUnregistered: true,
+        }
+      )
+
+      if (!FeatureFlag.isFeatureEnabled("caching") || !cachingModule) {
+        return await originalMethod.apply(this, args)
+      }
+
+      const resolvableKeys = [
+        "enable",
+        "key",
+        "tags",
+        "ttl",
+        "autoInvalidate",
+        "providers",
+      ]
+      const cacheOptions = {} as Parameters<typeof useCache>[1]
+
+      if (!options.key) {
+        options.key = await cachingModule.computeKey(
+          args.map((arg) => {
+            if (isObject(arg)) {
+              // Prevent any container, manager, transactionManager, etc from being included in the key
+              const {
+                container,
+                manager,
+                transactionManager,
+                __type,
+                ...rest
+              } = arg as any
+              return rest
+            }
+            return arg
+          })
+        )
+      }
+
+      const promises: Promise<any>[] = []
+      for (const key of resolvableKeys) {
+        if (typeof options[key] === "function") {
+          const res = options[key](args, cachingModule)
+          if (res instanceof Promise) {
+            promises.push(
+              res.then((value) => {
+                cacheOptions[key] = value
+              })
+            )
+          } else {
+            cacheOptions[key] = res
+          }
+        } else {
+          cacheOptions[key] = options[key]
+        }
+      }
+
+      await Promise.all(promises)
+
+      if (!cacheOptions.enable) {
+        return await originalMethod.apply(this, args)
```

Comment:
I think this can be moved to line 169

Reply:
I still need to process the enable option which is done line 203

Reply:
let me know what you think @carlos-r-l-rodrigues